scheduler/src/ratelimit.clj

```diff
@@ -0,0 +1,114 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns ratelimit
+  (:require [ratelimit.token-bucket-filter]
+            [clj-time.core]
+            [clj-time.coerce])
+  (:import (com.google.common.cache CacheBuilder Cache)
+           (java.util.concurrent TimeUnit)))
+
+(defprotocol RateLimiter
+  (earn-tokens! [this key]
+    "Account for any earned tokens for the requested key.")
+  (in-debt!? [this key]
+    "Boolean indicating if the rate limit is currently in debt. Earns tokens first, then checks to see if in debt.")
+  (spend-tokens! [this key tokens]
+    "Request this number of tokens. Does not block (or otherwise indicate we're in debt")
+  (time-until!? [this key]
+    "Time, in milliseconds, until this rate limiter is out of debt for this key. Earns tokens first.")
+  (is-enforcing? [this]
+    "Are we enforcing this policy?"))
+
+(defn now
```

Comment:
This name confusing as we expect `now` to return a date time object. Maybe choose another name, `clock` or `current-time-in-ms`?